src/MultiThreadedCaches.jl

```diff
@@ -0,0 +1,177 @@
+        # and potential thread safety violations due to Tasks migrating threads.
+        # NOTE that we only grab the lock if the key doesn't exist, so the mutex contention
+        # is not on the critical path for most accessses. :)
+        is_first_task = false
+        local future  # used only if the base_cache doesn't have the key
+        # We lock the mutex, but for only a short, *constant time* duration, to grab the
+        # future for this key, or to create the future if it doesn't exist.
+        @lock cache.base_cache_lock begin
+            value_or_miss = get(cache.base_cache, key, CACHE_MISS)
+            if value_or_miss !== CACHE_MISS
+                return value_or_miss::V
+            end
+            future = get!(cache.base_cache_futures, key) do
+                is_first_task = true
+                Channel{V}(1)
+            end
+        end
+        if is_first_task
+            v = try
+                func()
+            catch e
+                # In the case of an exception, we abort the current computation of this
+                # key/value pair, and throw the exception onto the future, so that all
+                # pending tasks will see the exeption as well.
+                #
+                # NOTE: we could also cache the exception and throw it from now on, but this
+                # would make interactive development difficult, since once you fix the
+                # error, you'd have to clear out your cache. So instead, we just rethrow the
+                # exception and don't cache anything, so that you can fix the exception and
+                # continue on. (This means that throwing exceptions remains expensive.)
+
+                # close(::Channel, ::Exception) requires an Exception object, so if the user
+                # threw a non-Exception, we convert it to one, here.
+                e isa Exception || (e = ErrorException("Non-exception object thrown during get!(): $e"))
+                close(future, e)
+                # As below, the future isn't needed after this returns (see below).
+                delete!(cache.base_cache_futures, key)
+                rethrow(e)
+            end
+            # Finally, lock again for a *constant time* to insert the computed value into
+            # the shared cache, so that we can free the Channel and future gets can read
+            # from the shared base_cache.
+            @lock cache.base_cache_lock begin
+                cache.base_cache[key] = v
+                # We no longer need the Future, since all future requests will see the key
+                # in the base_cache. (Other Tasks may still hold a reference, but it will
+                # be GC'd once they have all completed.)
+                delete!(cache.base_cache_futures, key)
+            end
+            # Return v to any other Tasks that were blocking on this key.
+            put!(future, v)
```

Comment:
will this send the value to all the tasks waiting on this channel, or just the first one? maybe it works because you're using `fetch`, not `take!`?

Reply:
yeah exactly, it's because i'm calling `fetch` not `take!`.

I'm not sure if it's exactly well-specified anywhere, but i've found that `Channel(1)` + `fetch()` makes for a Future. 👍